raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java

```diff
@@ -74,35 +74,50 @@ public void punctuate(long timestamp) {
     // always cancel the punctuator else it will get re-scheduled automatically
     cancellable.cancel();
 
-    Long emitTs = traceEmitTriggerStore.delete(key);
+    Long emitTs = traceEmitTriggerStore.get(key);
+    if (null == emitTs) {
+      /*
+        todo - debug why this happens .
+        Typically seen when punctuators are created via {@link RawSpansGroupingTransformer.restorePunctuators}
+       */
+      logger.warn("emitTs for key=[{}] is missing.", key);
+      return;
+    }
+
     if (emitTs <= timestamp) {
+      // we can emit this trace so just delete the entry for this 'key'
+      traceEmitTriggerStore.delete(key);
       // Implies that no new spans for the trace have arrived within the last 'groupingWindowTimeoutMs' interval
       // so the trace can be finalized and emitted
       ValueAndTimestamp<RawSpans> agg = inflightTraceStore.delete(key);
-      ByteBuffer traceId = null;
-      String customerId = null;
+      ByteBuffer traceId;
+      String customerId;
       if (!agg.value().getRawSpans().isEmpty()) {
         RawSpan firstSpan = agg.value().getRawSpans().get(0);
         traceId = firstSpan.getTraceId();
         customerId = firstSpan.getCustomerId();
-      }
-      List<RawSpan> rawSpanList = agg.value().getRawSpans();
-      Timestamps timestamps = trackEndToEndLatencyTimestamps(timestamp, agg);
-      StructuredTrace trace = StructuredTraceBuilder
-          .buildStructuredTraceFromRawSpans(rawSpanList, traceId, customerId,
-              timestamps);
 
-      logger.debug("Emit trace_id=[{}], spans_count=[{}]", key, rawSpanList.size());
-      context.forward(null, trace, outputTopicProducer);
+        List<RawSpan> rawSpanList = agg.value().getRawSpans();
+        Timestamps timestamps = trackEndToEndLatencyTimestamps(timestamp, agg);
+        StructuredTrace trace = StructuredTraceBuilder
+            .buildStructuredTraceFromRawSpans(rawSpanList, traceId, customerId,
+                timestamps);
+
+        logger.debug("Emit customer_id=[{}], trace_id=[{}], spans_count=[{}]",
+            customerId,
+            HexUtils.getHex(traceId),
+            rawSpanList.size());
+
+        context.forward(null, trace, outputTopicProducer);
+      }
```

Comment:
can this be true `agg.value().getRawSpans().isEmpty()`? I think this should not happen. we can put a log message in else block to check if you want.